src/context/AppProvider/AppContext.tsx

```diff
@@ -105,58 +103,51 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
+        const accountIds = Object.keys(result)
+
+        Object.assign(accountMetadataByAccountId, result)
+
         const { getAccount } = portfolioApi.endpoints
-        const opts = { forceRefetch: true }
-        // do *not* upsertOnFetch here - we need to check if the fetched account is empty
         const accountPromises = accountIds.map(accountId =>
-          dispatch(getAccount.initiate({ accountId }, opts)),
+          dispatch(getAccount.initiate({ accountId }, { forceRefetch: true })),
         )
+
         const accountResults = await Promise.allSettled(accountPromises)
-        /**
-         * because UTXO chains can have multiple accounts per number, we need to aggregate
-         * balance by chain id to see if we fetch the next by accountNumber
-         */
-        const balanceByChainId = accountResults.reduce<Record<ChainId, BN>>((acc, res, idx) => {
-          if (res.status === 'rejected') return acc
+
+        let chainIdsWithActivity: string[] = []
+        accountResults.forEach((res, idx) => {
+          if (res.status === 'rejected') return
+
           const { data: account } = res.value
-          if (!account) return acc
+          if (!account) return
+
           const accountId = accountIds[idx]
           const { chainId } = fromAccountId(accountId)
-          const accountBalance = Object.values(account.accountBalances.byId).reduce<BN>(
-            (acc, byAssetId) => {
-              Object.values(byAssetId).forEach(balance => (acc = acc.plus(bnOrZero(balance))))
-              return acc
-            },
-            bnOrZero(0),
-          )
-          acc[chainId] = bnOrZero(acc[chainId]).plus(accountBalance)
-          // don't upsert empty accounts past account 0
-          if (accountNumber > 0 && accountBalance.eq(0)) return acc
-          const accountMetadata = accountMetadataByAccountId[accountId]
-          const payload = { [accountId]: accountMetadata }
-          dispatch(portfolio.actions.upsertAccountMetadata(payload))
+          const { hasActivity } = account.accounts.byId[accountId]
+
+          // don't add accounts with no activity past account 0
+          if (accountNumber > 0 && !hasActivity) return delete accountMetadataByAccountId[accountId]
+
+          chainIdsWithActivity = Array.from(new Set([...chainIdsWithActivity, chainId]))
           dispatch(portfolio.actions.upsertPortfolio(account))
-          return acc
-        }, {})
-
-        /**
-         * if the balance for all accounts for the current chainId and accountNumber
-         * is zero, we've exhausted that chain, don't fetch more of them
-         */
-        Object.entries(balanceByChainId).forEach(([chainId, balance]) => {
-          if (balance.eq(0)) pull(chainIds, chainId) // pull mutates chainIds, but we want to
         })
+
+        chainIds = chainIdsWithActivity
       }
+
+      dispatch(portfolio.actions.upsertAccountMetadata(accountMetadataByAccountId))
```

Comment:
only upsert account metadata once at the very end to ensure the `getAllTxHistory` effect is not triggered multiple times for `requestedAccountIds`